adhocteam/Head-Start-TTADP
# Import Activity Reports from smartsheets

**Description of change**

This adds a script to import activity reports from CSV files. If you have a collection of `R\d{1,2}*.csv` CSV files (in the format exported from smartsheets), you should be able to upload those to s3 (minio) and `yarn import:reports:local --file file.csv --region 14` and import CSV data. Check out `src/tools/importActivityReports.js` for the bulk of the work.

There is a JSONB field, `imported`, for storing information from the parsed CSV, which would allow for future improvements to imported data. The import process attempts to convert rows into records, with various caveats:

- The importer is set to skip rows that don't have a report id, since those are likely blank rows.
- The import does not many address relational fields. For example, User records won't be created prior to import, and there is no guarantee that the emails from smartsheets will match up.
- Grant/Grantee records. If a Grant with a matching name can be found, an `ActivityRecipient` will be created.
- `{ validate: false }`, because `checkRequiredForSubmission` checks for values that may not exist during import
- Records that started with *R14* were given a `legacyId` that uses the actual region id (taken from the filename) rather than the test region 14. The `regionId` is taken from `--region` cli option and used to correct the region in the identifier. You can still get the old value via `"imported"->'reportId'`.
- Depending on how you convert from Excel to csv (save as from Excel, use `csvkit`), date fields in the output may differ. We try to address that in `coerceDate`, but YMMV.

**How to test**

- Get your hands on some Smartsheet files, convert to csv, upload them to minio.
- `yarn db:migrate` to remove some "not null" requirements, create the "imported" JSONB column
- run `yarn import:reports:local --file R1ActivityReports.csv --region 1` or similar.
- Check the db: `SELECT count(id) FROM "ActivityReports" WHERE "legacyId" IS NOT NULL`, for example

**Issue(s)**
* https://github.com/HHS/Head-Start-TTADP/issues/215

**Checklist**

    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified
